src/cpu.c

```diff
@@ -324,168 +349,228 @@ static int init(void) {
-                           gauge_t rates[static COLLECTD_CPU_STATE_MAX]) {
-  metric_family_t fam = {
-      .name = "system.cpu.utilization",
-      .type = METRIC_TYPE_GAUGE,
-  };
+    value_t v = {0};
+    int status = rate_to_value(&v, 1000000.0 * state_ratio[s], &us->to_count,
+                               DS_TYPE_DERIVE, u->time);
+    if (status == 0) {
+      us->count = v.derive;
+      us->has_count = true;
+    }
+  }
 
-  metric_t m = {0};
-  if (cpu_num == -1) {
-    metric_label_set(&m, "cpu", "total");
-  } else {
-    char cpu_num_str[32];
-    ssnprintf(cpu_num_str, sizeof(cpu_num_str), "%d", cpu_num);
-    metric_label_set(&m, "cpu", cpu_num_str);
+  u->finalized = true;
+}
+
+static void usage_reset(usage_t *u) {
+  if (u == NULL) {
+    return;
   }
+  free(u->states);
+  memset(u, 0, sizeof(*u));
+}
 
-  gauge_t sum = rates[COLLECTD_CPU_STATE_ACTIVE];
-  RATE_ADD(sum, rates[COLLECTD_CPU_STATE_IDLE]);
+static gauge_t usage_rate(usage_t *u, size_t cpu, state_t state) {
+  usage_finalize(u);
 
-  if (!report_by_state) {
-    m.value.gauge = 100.0 * rates[COLLECTD_CPU_STATE_ACTIVE] / sum;
-    metric_label_set(&m, "state", cpu_state_names[COLLECTD_CPU_STATE_ACTIVE]);
-    metric_family_metric_append(&fam, m);
+  usage_state_t us;
+  if (cpu == CPU_ALL) {
+    us = u->global[state];
   } else {
-    for (size_t state = 0; state < COLLECTD_CPU_STATE_ACTIVE; state++) {
-      m.value.gauge = 100.0 * rates[state] / sum;
-      metric_label_set(&m, "state", cpu_state_names[state]);
-      metric_family_metric_append(&fam, m);
+    size_t index = (cpu * STATE_MAX) + state;
+    if (index >= u->states_num) {
+      return -1;
     }
+    us = u->states[index];
   }
 
-  int status = plugin_dispatch_metric_family(&fam);
-  if (status != 0) {
-    ERROR("cpu plugin: plugin_dispatch_metric_family failed: %s",
-          STRERROR(status));
+  return us.has_rate ? us.rate : NAN;
+}
+
+static gauge_t usage_ratio(usage_t *u, size_t cpu, state_t state) {
+  usage_finalize(u);
+
+  gauge_t global_rate =
+      usage_rate(u, CPU_ALL, STATE_ACTIVE) + usage_rate(u, CPU_ALL, STATE_IDLE);
+  return usage_rate(u, cpu, state) / global_rate;
+}
+
+static derive_t usage_count(usage_t *u, size_t cpu, state_t state) {
+  usage_finalize(u);
+
+  usage_state_t us;
+  if (cpu == CPU_ALL) {
+    us = u->global[state];
+  } else {
+    size_t index = (cpu * STATE_MAX) + state;
+    if (index >= u->states_num) {
+      return -1;
+    }
+    us = u->states[index];
   }
 
-  metric_reset(&m);
-  metric_family_metric_reset(&fam);
-  return;
-} /* }}} void cpu_commit_one */
+  return us.has_count ? us.count : -1;
+}
 
 /* Commits the number of cores */
 static void cpu_commit_num_cpu(gauge_t value) /* {{{ */
 {
   metric_family_t fam = {
-      .name = "cpu_count",
+      .name = "system.cpu.logical.count",
+      .help = "Reports the number of logical (virtual) processor cores created "
+              "by the operating system to manage multitasking",
```

Comment:
Does this mean hyper-threading or something else?  And how those would be OS-created instead of HW feature, or is this about VMs?  In any case, I think what "creates" the cores not to be relevant:
```suggestion
      .help = "Reports the number of logical (virtual) processor cores",
```

Reply:
Done.